File: pa2/List.c
//
// Created by E on 10/3/16.
//



#include <stdlib.h>
#include "List.h"


typedef struct NodeObj {
    int data;
    struct NodeObj *next;
    struct NodeObj *prev;
} NodeObj;

// private Node type
typedef NodeObj *Node;


typedef struct ListObject {
    Node front;
    Node back;
    Node cursor;
    int index;
    int length;

} ListObject;

void freeNode(Node *n) {
    if (n != NULL && *n != NULL) {
        free(*n);
        *n = NULL;
    }
}

Node newNode(int d) {
    Node x = malloc(sizeof(NodeObj));
    x->data = d;
    x->prev = NULL;
    x->next = NULL;
    return (x);
}


List newList(void) {
    List temp = malloc(sizeof(ListObject));
    temp->front = temp->back = temp->cursor = NULL;
    temp->index = -1;
    temp->length = 0;
    return temp;
}


void freeList(List* pL) {
    if(pL != NULL && *pL != NULL) {
        Node x = (*pL)->front;
        while(x != NULL) {
            Node temp = x;
            x = x->next;
            free(temp);
        }
        free(*pL);
        *pL = NULL;
    }
}

// Access functions -----------------------------------------------------------
int length(List L) {
    return L->length;
}

int index(List L) {
    return L->index;
}

int front(List L) {
    if (L->length > 0) {
        return L->front->data;
    }
    return -1;
}

int back(List L) {
    if (L->length > 0) {
        return L->back->data;
    }
    return -1;
}

int get(List L) {
    if (L->length > 0) {
        return L->cursor->data;
    }
    return -1;
}

int equals(List A, List B) {
    Node firstList = A->front;
    Node secondList = B->front;

    if (A->length != B->length) {
        return 0;
    }
    if (A->back->data != B->back->data) {
        return 0;
    }
    while (firstList->next != NULL) {
        if (firstList->data != secondList->data) {
            return 0;
        }

        firstList = firstList->next;
        secondList = secondList->next;
    }
    return 1;
}

// Manipulation procedures ----------------------------------------------------
void clear(List L) {
    if(L == NULL) {
        printf("EMPTY LIST");
        return;
    }
    Node temp = L->front;
    while(temp != NULL) {
        Node x = temp;
        temp = temp->next;
        free(x);
    }
    L->front = NULL;
    L->back = NULL;
    L->cursor = NULL;
    L->length = 0;
    L->index = -1;

}

void moveFront(List L) {
    if (L->length > 0) {
        L->cursor = L->front;
        L->index = 0;
    }
}

void moveBack(List L) {
    if (L->length > 0) {
        L->cursor = L->back;
        L->index = L->length - 1;
    }
}

void movePrev(List L) {

    if (L->cursor != NULL && L->cursor != L->front) {
        L->cursor = L->cursor->prev;
        --L->index;

    } else if (L->cursor != NULL && L->cursor == L->front) {
        L->cursor = NULL;
        L->index = -1;
    } else {
        return;
    }
}

void moveNext(List L) {

    if (L->cursor != NULL && L->cursor != L->back) {
        L->cursor = L->cursor->next;
        ++L->index;
    } else if (L->cursor != NULL && L->cursor == L->back) {
        L->cursor = NULL;
        L->index = -1;
    } else {
        return;
    }
}

void prepend(List L, int data) {
    if (L->length > 0) {
        Node newNd = newNode(data);
        (newNd)->next = L->front;
        L->front->prev = (newNd);
        L->front = (newNd);
        ++L->length;
        ++L->index;

    } else {
        Node newNd = newNode(data);
        L->front = newNd;
        L->back = newNd;
        ++L->length;
        ++L->index;
    }

}

void append(List L, int data) {
    if (L->length > 0) {
        Node newNd = newNode(data);
        (newNd)->prev = L->back;
        L->back->next = (newNd);
        L->back = (newNd);
        ++L->length;

    } else {
        Node newNd = newNode(data);
        L->front = (newNd);
        L->back = (newNd);
        ++L->length;

    }
}

void insertBefore(List L, int data) {
    if (L->length > 0 && L->index >= 0) {
        Node newNd = newNode(data);
        (newNd)->data = data;
        (newNd)->prev = L->cursor->prev;
        (newNd)->next = L->cursor;
        L->cursor->prev->next = (newNd);
        L->cursor->prev = (newNd);
        ++L->index;
        ++L->length;
    }

}

void insertAfter(List L, int data) {
    if (L->length > 0 && L->index >= 0) {
        Node newNd = newNode(data);
        (newNd)->data = data;
        (newNd)->next = L->cursor->next;
        (newNd)->prev = L->cursor;
        L->cursor->next =  (newNd);
        ++L->length;

    }
}

void deleteFront(List L) {
    Node x = *&(L->front);

    printf(" deleteig %d", x->data);
    fflush(stdout);

    if (L->length > 1) {
        L->front = L->front->next;
        freeNode(&x);
    }
    else
    {
        free(&(L->front));
        L->front = L->back = NULL;
    }
    --L->length;

    printf(" the new length is %d \n", L->length);
    fflush(stdout);
}

void deleteBack(List L) {
    Node x = *&(L->back);

    if (L->length > 1) {
        L->back = L->back->prev;
        freeNode(&x);
    }
    else
    {
        free(&(L->back));
        L->front = L->back = NULL;
    }
    --L->length;
}

void delete(List L) {
    if (L->length > 0 && L->index >= 0) {
        if(L->cursor == L->back) {
            deleteBack(L);
        } else if(L->cursor == L->front) {
            deleteFront(L);
        } else {
            Node temp = L->cursor;
            L->cursor->next->prev = L->cursor->prev;
            L->cursor->prev->next = L->cursor->next;
            freeNode(&temp);
            L->index = -1;
            L->cursor = NULL;
            --L->length;
        }
    }
}

// Other operations -----------------------------------------------------------
void printList(FILE *out, List L) {

    Node temp = L->front;


    if(out == NULL) {

        for (int i = 0; i < L->length; i++) {
            printf(" %d ", temp->data);
            temp = temp->next;
        }
    }
    else
    {
        for (int i = 0; i < L->length; i++) {
            fprintf(out," %d ", temp->data);
            temp = temp->next;
        }
    }
}

List copyList(List L) {

    List temp = newList();
    Node tempNode = L->front;
    while (tempNode->next != NULL)
    {
        append(temp,tempNode->data);
        tempNode = tempNode->next;
    }

    append(temp,tempNode->data);

    return temp;
}

File: pa3/List.java
/**
 * Created by eyad murshid on 9/26/16.
 */
public class List {

    //Node inner class
    private class Node {

        Object data;
        Node next;
        Node prev;

        Node(Object data) {
            this.data = data;
            next = null;
            prev = null;
        }

        Node(Object data, Node nextNode, Node prevNode) {
            this.data = data;
            next = nextNode;
            prev = prevNode;
        }

        // toString():  overrides Object's toString() method
        public String toString() {
            return String.valueOf(data);
        }

    }

    // Fields
    private Node front;
    private Node back;
    private Node cursor;
    private int index;
    private int length;

    // Constructor
    List() {
        length = 0;
        front = null;
        cursor = null;
        back = null;
        index = -1;

    }

    // Returns the number of elements in this List.
    int length() {
        return length;
    }

    // Returns the number of elements in this List.
    // otherwise returns -1.
    int index() {
        return index;
    }

    // Returns front element. Pre: length()>0
    Object front() {
        if (length > 0) {
            return front.data;
        }
        return -1;
    }

    // Returns back element. Pre: length()>0
    Object back() {
        if (length > 0) {
            return back.data;
        }
        return -1;
    }

    // Returns cursor element. Pre: length()>0, index()>=0
    Object get() {
        if (length() == 0) {
            return null;
        } else if (index == -1) {
            return null;
        } else {
            return cursor.data;
        }
    }

    // Resets this List to its original empty state
    void clear() {
        cursor = front;
        if (cursor != null) {
            while (cursor.next != null) {
                Node temp = cursor;
                cursor = cursor.next;
                temp.next = null;
                temp.prev = null;
            }
            cursor = null;
            front = null;
            back = null;
            length = 0;
            index = -1;
        }

    }

    // If List is non-empty, places the cursor under the front element,
    // otherwise does nothing.
    void moveFront() {
        if (length > 0) {
            cursor = front;
            index = 0;
        }
    }

    // If List is non-empty, places the cursor under the back element,
    // otherwise does nothing.
    void moveBack() {
        if (length > 0) {
            cursor = back;
            index = length - 1;
        }
    }

    // If cursor is defined and not at front, moves cursor one step toward
    // front of this List, if cursor is defined and at front, cursor becomes
    // undefined, if cursor is undefined does nothing
    void movePrev() {
        if (cursor != null && !cursor.equals(front)) {
            cursor = cursor.prev;
            index--;
        } else if (cursor != null && cursor.equals(front)) {
            cursor = null;
            index = -1;
        } else
            return;

    }

    // If cursor is defined and not at back, moves cursor one step toward
    // back of this List, if cursor is defined and at back, cursor becomes
    // undefined, if cursor is undefined does nothing.
    void moveNext() {

        if (cursor != null && !cursor.equals(back)) {
            cursor = cursor.next;
            index++;
        } else if (cursor != null && cursor.equals(back)) {
            cursor = null;
            index = -1;
        } else
            return;
    }

    // Insert new element into this List. If List is non-empty,
    // insertion takes place before front element.
    void prepend(Object data) {
        if (length > 0) {
            Node newFront = new Node(data, front, null);
            front.prev = newFront;
            front = newFront;
            length++;
            index++;
        } else {
            Node newFront = new Node(data);
            front = newFront;
            back = newFront;
            length++;
            index++;
        }
    }

    // Insert new element into this List. If List is non-empty,
    // insertion takes place after back element.
    void append(Object data) {
        if (length > 0) {
            Node newFront = new Node(data, null, back);
            back.next = newFront;
            back = newFront;
            length++;
        } else {
            Node newFront = new Node(data);
            front = newFront;
            back = newFront;
            length++;
        }
    }

    // Insert new element before cursor.
    // Pre: length()>0, index()>=0
    void insertBefore(Object data) {
        if (length() <= 0) {
            return;
        } else if (index() < 0) {
            return;
        }
        Node newNode = new Node(data);
        if (index() == 0) {
            prepend(data);
        } else {
            newNode.next = cursor;
            newNode.prev = cursor.prev;
            cursor.prev.next = newNode;
            cursor.prev = newNode;
            index++;
            length++;
        }
    }

    // Inserts new element after cursor.
    // Pre: length()>0, index()>=0
    void insertAfter(Object data) {
        if (length() <= 0) {
            return;
        } else if (index() < 0) {
            return;
        }
        Node newNode = new Node(data);
        if (index() == length() - 1) {
            append(data);
        } else {
            newNode.prev = cursor;
            newNode.next = cursor.next;
            cursor.next.prev = newNode;
            cursor.next = newNode;
            length++;
        }

    }

    void advanceTo(int i) {
        if (i == index) return;
        if (i >= length() || i < 0) {
            cursor = null;
            index = -1;
            return;
        }
        if (i == 0)
        {
            cursor = front;
        }
        if (i == length() - 1)
        {
            cursor = back;
        }
        else {
            int splitter = 0;
            int first = length() - i;
            int second = i;
            int third = Math.abs(index - i);
            int min = first;
            if (min > second)
            {
                min = second;
            }
            if (min > third)
            {
                min = third;
            }
            if (min == first) {
                cursor = back;
                index = length() - 1;
                splitter = 0;
            } else if (min == second) {
                cursor = front;
                index = 0;
                splitter = 1;
            } else if (min == third) {
                if (i > index) splitter = 1;
            }
            for (int j = 0; j < i; j++) {
                if (splitter == 1) moveNext();
                else movePrev();
            }
        }
        index = i;
    }

    // Deletes the front element. Pre: length()>0
    void deleteFront() {
        if (length > 0) {
            front = front.next;
            if (front != null) {
                front.prev = null;
            }
            length--;
        } else {
            front = null;
        }
    }

    // Deletes the back element. Pre: length()>0
    void deleteBack() {
        if (length > 0) {
            back = back.prev;
            if (back != null) {
                back.next = null;
            }
            length--;
        } else {
            back = null;
        }
    }

    // Deletes cursor element, making cursor undefined.
    // Pre: length()>0, index()>=0
    void delete() {
        if (length > 0 && index >= 0) {
            cursor.next.prev = cursor.prev;
            cursor.prev.next = cursor.next;
            cursor = null;
            index = -1;
            length--;
        }
    }

    public boolean equals(Object x) {
        List L;
        if (x instanceof List) {
            L = (List) x;
            if (this.length == L.length) {
                Node firstList = this.front;
                Node secondList = L.front;
                while (firstList != null && secondList != null) {
                    if (firstList.data.equals(secondList.data)) {
                        firstList = firstList.next;
                        secondList = secondList.next;
                    } else return false;
                }
                return true;
            } else {
                return false;
            }
        } else return false;
    }


    // Overrides Object's toString method. Returns a String
    // representation of this List consisting of a space
// separated sequence of integers, with front on left.
    public String toString() {
        String str = "";
        for (Node N = front; N != null; N = N.next) {
            str += N.toString() + " ";
        }
        return str;
    }
}
